### PR TITLE
hotpatch: implement /hotpatch disable

### PR DESCRIPTION
**Pull request by panpawn**
Allows developers to turn off the use of hotpatch.

### CODE REVIEW

**Comment by Slayer95** [comment]
I don't think we need the ``enable`` variant.

**Comment by panpawn** [comment]
Do you think it could still be useful, though?  I mean, you could use eval to turn it off obviously, I guess - if need be.

**Comment by Slayer95** [comment]
As a general rule, eval-patching ``/hotpatch`` is a terrible practice. Hence why having a predictable command for disabling it is the best way to change its behavior.

Under normal circumstances, a server whose internal state has come into an un-hotpatchable status will not become patchable again.

The only way to fix it would be with ``eval``. Someone running such eval commands successfully will probably be able to deactive the hotpatch blocker flag as well. If that condition has not been met, however, enabling hotpatch is a time bomb.

**Comment by panpawn** [comment]
Updated

**Comment by panpawn** [comment]
(specified until the next server restart as well)

**Comment by Zarel** [comment]
Honestly, though, hotpatching the server is fine as long as you don't run `/updateserver` or `git pull` or something...

**Comment by Slayer95** [comment]
So basically hotpatch it to the same version? !_!

It can be done just before disabling it...

**Comment by Zarel** [comment]
I mean, it's possible to fork the code...

**Comment by Zarel** [comment]
Like, I can still make bugfixes to commands on the server and commit those, without pulling in the `RoomGame` refactor.

**Comment by Slayer95** [comment]
Then you wouldn't disable ``/hotpatch`` until after pulling the refactor I guess.

File: commands.js [hunk 1]
```diff
@@ -1792,6 +1797,10 @@ exports.commands = {
 			} catch (e) {
 				return this.errorReply("Something failed while trying to hotpatch learnsets: \n" + e.stack);
 			}
+		} else if (target === 'disable') {
+			if (CommandParser.hotpatchLock) return this.errorReply("Hotpatch is already disabled.");
+			CommandParser.hotpatchLock = true;
```
Comment (Zarel) [inline commands.js]: Might as well allow specifying the reason here...

**Comment by panpawn** [comment]
Will fix travis momentarily

**Comment by panpawn** [comment]
How does this look?

File: commands.js [hunk 2]
```diff
@@ -1792,6 +1797,12 @@ exports.commands = {
 			} catch (e) {
 				return this.errorReply("Something failed while trying to hotpatch learnsets: \n" + e.stack);
 			}
+		} else if (target.indexOf('disable') >= 0) {
```
Comment (Zarel) [inline commands.js]: `.startsWith`

File: commands.js [hunk 3]
```diff
@@ -1792,6 +1797,12 @@ exports.commands = {
...
+			if (CommandParser.hotpatchLock) return this.errorReply("Hotpatch is already disabled.");
```
Comment (Zarel) [inline commands.js]: `Config` might be a better global than `CommandParser` for this.

**Comment by panpawn** [comment]
Changed it to use ``Config`` and used ``.startsWith()``

**Comment by Slayer95** [comment]
If in ``Config``, it can be reset on editions for ``watchconfig = true`` though...

**Comment by Zarel** [comment]
`Monitor`? That's a global that can't get hotpatched...

**Comment by panpawn** [comment]
Changed it to use ``Monitor``

**Comment by panpawn** [comment]
Does anyone foresee any issues with using ``Monitor``?  (BTW, that was a good catch with using ``Config``, Slayer)

**Pull request merged**